tensorflow/compiler/mlir/tosa/transforms/legalize_utils.h

```diff
@@ -119,6 +121,54 @@ llvm::Optional<Value> getConstTensor(PatternRewriter& rewriter, Operation* op,
 // Check if scale32 mode is used for given output_element_type
 bool isScale32(mlir::quant::UniformQuantizedType output_element_type);
 
+// Creates a TOSA operation and performs shape inference on the individual
+// op. This allows shape inference during the TFLite to TOSA lowering.
+template <typename TosaOp, typename... Args>
+TosaOp CreateOpAndInfer(PatternRewriter& rewriter, Location loc, Type result_ty,
+                        Args&&... args) {
+  auto op = rewriter.create<TosaOp>(loc, result_ty, args...);
+
+  InferShapedTypeOpInterface shapeInterface =
+      dyn_cast<InferShapedTypeOpInterface>(op.getOperation());
+  if (!shapeInterface) return op;
+
+  SmallVector<ShapedTypeComponents> returnedShapes;
+  if (shapeInterface
+          .inferReturnTypeComponents(op.getContext(), op.getLoc(),
+                                     op->getOperands(), op->getAttrDictionary(),
+                                     op->getRegions(), returnedShapes)
+          .failed())
+    return op;
+
+  auto result = op->getResult(0);
```

Comment:
Can you please provide some justificaiton/example of when inferReturnTypeComponents is not sufficient and must be augmented with the DFA? It isn't clear to me.

Reply:
Explanation added.

Basically tosa.rescale requires output type information not specified on the operation. Using the existing output type gives us the full type (not possible via pure shape inference). It may make sense to add a TypeAttr on rescales to avoid this foo. It would also allow us to more easily define TOSA models as a list of operations.